NDTensors/src/lib/Sectors/src/abstractcategory.jl

```diff
@@ -1,45 +1,114 @@
+block_dimensions(g::AbstractUnitRange) = block_dimensions(SymmetryStyle(g), g)
+block_dimensions(::AbelianStyle, g) = unlabel.(blocklengths(g))
+function block_dimensions(::NotAbelianStyle, g)
+  return quantum_dimension.(blocklabels(g)) .* blocklengths(g)
+end
+
+quantum_dimension(x) = quantum_dimension(SymmetryStyle(x), x)
+
+function quantum_dimension(::NotAbelianStyle, c::AbstractCategory)
+  return error("method `quantum_dimension` not defined for type $(typeof(c))")
 end
 
+quantum_dimension(::AbelianStyle, ::AbstractCategory) = 1
+quantum_dimension(::AbelianStyle, g::AbstractUnitRange) = length(g)
+quantum_dimension(::NotAbelianStyle, g::AbstractUnitRange) = sum(block_dimensions(g))
+
+# ===============================  Fusion rule interface  ==================================
 ⊗(c1::AbstractCategory, c2::AbstractCategory) = fusion_rule(c1, c2)
 
-⊕(c1::AbstractCategory, c2::AbstractCategory) = [c1, c2]
-⊕(cs::Vector{<:AbstractCategory}, c::AbstractCategory) = [cs; c]
-⊕(c::AbstractCategory, cs::Vector{<:AbstractCategory}) = [c; cs]
+function fusion_rule(c1::AbstractCategory, c2::AbstractCategory)
+  return fusion_rule(combine_styles(SymmetryStyle(c1), SymmetryStyle(c2)), c1, c2)
+end
+
+function fusion_rule(::NotAbelianStyle, c1::C, c2::C) where {C<:AbstractCategory}
+  degen, labels = label_fusion_rule(C, category_label(c1), category_label(c2))
+  return gradedrange(labelled.(degen, C.(labels)))
+end
+
+# abelian case: return Category
+function fusion_rule(::AbelianStyle, c1::C, c2::C) where {C<:AbstractCategory}
+  return C(label_fusion_rule(C, category_label(c1), category_label(c2)))
+end
+
+function label_fusion_rule(category_type::Type{<:AbstractCategory}, ::Any, ::Any)
+  return error("`label_fusion_rule` not defined for type $(category_type).")
+end
+
+# ================================  GradedAxes interface  ==================================
+# tensor_product interface
+function GradedAxes.fuse_blocklengths(
+  l1::LabelledInteger{<:Integer,<:AbstractCategory},
+  l2::LabelledInteger{<:Integer,<:AbstractCategory},
+)
+  return fuse_blocklengths(combine_styles(SymmetryStyle(l1), SymmetryStyle(l2)), l1, l2)
+end
 
-function Base.show(io::IO, cs::Vector{<:AbstractCategory})
-  (length(cs) <= 1) && print(io, "[")
-  symbol = ""
-  for c in cs
-    print(io, symbol, c)
-    symbol = " ⊕ "
-  end
-  (length(cs) <= 1) && print(io, "]")
-  return nothing
+function GradedAxes.fuse_blocklengths(
+  ::NotAbelianStyle, l1::LabelledInteger, l2::LabelledInteger
+)
+  fused = label(l1) ⊗ label(l2)
+  v = labelled.(l1 * l2 .* blocklengths(fused), blocklabels(fused))
+  return gradedrange(v)
 end
 
-function trivial(category_type::Type{<:AbstractCategory})
-  return error("`trivial` not defined for type $(category_type).")
+function GradedAxes.fuse_blocklengths(
+  ::AbelianStyle, l1::LabelledInteger, l2::LabelledInteger
+)
+  fused = label(l1) ⊗ label(l2)
+  return labelled(l1 * l2, fused)
```

Comment:
@ogauthe would it simplify `tensor_product` to return a graded unit range here, i.e. `gradedrange([labelled(l1 * l2, fused)])`? I'm not sure where else this is used or if that has other consequences, I guess I was the one who originally wrote `fuse_blocklengths` but I can't remember the thought process behind its design so we should feel free to change it if it simplifies the tensor product code.

Reply:
That would not be enough, as `tensor_product` would still need an abelian-like behavior for e.g. string labels. `tensor_product` needs to be specialized for `AbstractCategory` label by defining
```julia
GradedAxes.tensor_product(
  a1::AbstractUnitRange{<:LabelledInteger{<:Integer,<:AbstractCategory}},
  a2::AbstractUnitRange{<:LabelledInteger{<:Integer,<:AbstractCategory}},
)
``` 
This is a work in progress.

The alternative is to make the default version in `GradedAxes` to behave in a non-abelian way and impose the generic `fuse_blocklength` to return a `Vector{<:Integer}`.

Reply:
I'm not following the details well enough to understand your comment and the tradeoffs between the different design choices. It sounds like we would have to discuss this in more detail to reach a conclusion about how to proceed, though feel free to try to refactor the code in a way that would make it simpler if you have an idea.

Reply:
I finally went for the alternative, imposing a non-abelian like interface in `GradedAxes` where `fuse_blocklengths` returns a `BlockedUnitRange`. The code ends up being pretty simple.

Reply:
That's basically what I was advocating for, but maybe I didn't describe it clearly enough. I'll take a closer look, but at first glance it looks good, nice to see you were able to remove `flatten_maybe_nested`.

Reply:
I think the current version looks good to me, thanks!